torch_dftd/functions/edge_extraction.py

```diff
@@ -4,9 +4,73 @@
 import torch
 from ase.neighborlist import primitive_neighbor_list
 from ase.units import Bohr
+from pymatgen.core import Structure
 from torch import Tensor
 
 
+def calc_neighbor_by_ase(
+    pos: Tensor, pbc: Tensor, cell: Tensor, cutoff: float
+) -> Tuple[Tensor, Tensor]:
+    idx_i, idx_j, S = primitive_neighbor_list(
+        "ijS",
+        pbc.detach().cpu().numpy(),
+        cell.detach().cpu().numpy(),
+        pos.detach().cpu().numpy(),
+        cutoff,
+    )
+    edge_index = torch.tensor(np.stack([idx_i, idx_j], axis=0), device=pos.device)
+    # convert int64 -> pos.dtype (float)
+    S = torch.tensor(S, dtype=pos.dtype, device=pos.device)
+    return edge_index, S
+
+
+def calc_neighbor_by_pymatgen(
+    pos: Tensor, cell: Tensor, pbc: Tensor, cutoff: float
+) -> Tuple[Tensor, Tensor]:
+    """calculate neighbor nodes in pbc condition.
+
+    Implementation referred from https://github.com/Open-Catalyst-Project/ocp/blob/a5634ee4f0dc4a874752ab8d3117492ce83261ac/ocpmodels/preprocessing/atoms_to_graphs.py#L76
+    under MIT license.
+
+    Args:
+        pos (Tensor):
+        cell (Tensor):
+        pbc (Tensor): periodic boundary condition.
+        cutoff (float): cutoff distance to find neighbor
+
+    Returns:
+        edge_index (Tensor): (2, n_edges) indices of edge, src -> dst.
+        S (Tensor): (n_edges, 3) shift tensor
+    """  # NOQA
+    if not torch.all(pbc):
+        raise NotImplementedError(f"pbc {pbc} must be true for all axis!")
+
+    positions = pos.detach().cpu().numpy().copy()
+    lattice = cell.detach().cpu().numpy().copy()
+    n_atoms = positions.shape[0]
+    symbols = np.ones(n_atoms)  # Dummy symbols to create `Structure`...
+
+    struct = Structure(lattice, symbols, positions, coords_are_cartesian=True)
+    c_index, n_index, offsets, n_distance = struct.get_neighbor_list(
+        r=cutoff,
+        numerical_tol=0,
+        exclude_self=True,
+    )
+    edge_index = torch.tensor(
+        np.stack([c_index, n_index], axis=0), dtype=torch.long, device=pos.device
+    )
+    edge_distances = torch.tensor(n_distance, dtype=torch.float32, device=pos.device)
+    cell_offsets = torch.tensor(offsets, dtype=torch.long, device=pos.device)
+
+    # remove distances smaller than a tolerance ~ 0. The small tolerance is
+    # needed to correct for pymatgen's neighbor_list returning self atoms
+    # in a few edge cases.
+    nonzero = torch.where(edge_distances >= 1e-8)[0]
```

Comment:
Maybe it is good to enforce that `dtype` of `positions` is always `float64` because of this line.

Reply:
I am not sure where this small tolerance will come from, but it seems some kind of subtraction seems to occur in somewhere, which will reduce the numerical accuracy. 
For example, the difference between 1.0 and next to 1.0 in fp32 is log_10{2^-24}, which is around 6.0e-8 in decimal value. So I feel a bit dangerous about this value.